### ISSUE TITLE
jschart outlier mitigation

### ISSUE THREAD

**Issue by k-rister** [distributed-system-analysis/pbench#317]
Sometimes a tool produces a significant outlier, a value that is several orders of magnitude different than every over value that was produced.  This results in an ugly, indecipherable chart.  The jschart package should provide a mitigation technique that allows the outlier's influence to be minimized.

![disk-outlier](https://cloud.githubusercontent.com/assets/15911116/17570896/d4fd5460-5f13-11e6-8b8e-e4d8d42020a8.png)

**Comment by k-rister** [comment]
Within jschart the data sets are analyzed to determine their minimum and maximum values for the X and Y axes.  These values are used to determine the domains for the X and Y axes -- the minimum value and the maximum value across all datasets are used as the axis domain.  The axis domain sets the bounds within which the chart can zoomed -- meaning that an axis domain range is the maximum the axis can be zoomed out.

The ability for the axis domain's to be modified already exists and is used if a dataset is hidden -- meaning that when a dataset is hidden the axis domain's are recalculated with the hidden dataset(s) being ignored.  If the hidden dataset contains one of the problematic outliers, then when the domains are recalculated the outlier's affects will be removed.  In the case of the example provided with this issue, hiding 7 datasets results in a much different appearing chart:

![disk-outlier-mitigated](https://cloud.githubusercontent.com/assets/15911116/17571158/a73c3f40-5f14-11e6-8443-50750f65c808.png)

The problem with this "solution" is that in order to remove the influence of an outlier, all values of a dataset must be hidden.

By creating an interface which the domain can be set by the user, outlier mitigation will exist **AND** the user will be further empowered to manipulate the viewport beyond what is currently possible.  In addition to solving the outlier issue, allowing the viewport to be manipulated in this manner could prove valuable in situations where the valid domain is very broad.  In this situation zooming in with the slider controls becomes difficult/impossible with a sufficiently high zoom factor.

By resolving the problem in this manner jschart will avoid passing judgment on the magnitude of individual datapoints and continue it's value agnostic approach to data visualization (one of my personal goals for the code is to avoid it evaluating the values -- it should simply process and present them to the user).

**Comment by k-rister** [comment]
I've done some prototyping to try and generate an example of what this would look like if it were implemented.  When the Y axis domain maximum is reduced to a sane value (13,000 in this case) the datasets with outliers are no longer rendered.  So far I don't believe that this is something that I have control over.

The default Y axis domain is [0, 18815678955183743000].  If I start to reduce the maximum Y axis domain value, eventually the datasets with outliers stop rendering properly.  Once the maximum value drops below 12850536692227 the line starts to only partially render and as the maximum domain value keeps dropping eventually the line is not rendered at all.

Y domain set to [0,12850536692227]:

![ok-y-domain](https://cloud.githubusercontent.com/assets/15911116/17604971/d361bc70-5fdd-11e6-88b7-6406cad4c162.png)

Y domain set to [0,12850536692226];

![broken-y-domain](https://cloud.githubusercontent.com/assets/15911116/17604981/dd44d40c-5fdd-11e6-9a51-78a28391c899.png)

My current hypothesis is that it's a rendering artifact caused by the large coordinate delta.

If I set the Y domain to [0,13000] I get exactly what I want, except for the fact that the datasets with outliers have disappeared without being hidden:

![test-y-domain](https://cloud.githubusercontent.com/assets/15911116/17605263/260d64fa-5fdf-11e6-8073-71add9818a2a.png)

I'm still trying to figure out if there is a way to resolve this without abandoning this approach entirely.

**Comment by k-rister** [comment]
Usually my testing is done in Chrome, but I did a quick sanity check in Firefox and it does different, but equally broken, stuff when I manually manipulate the Y axis domain.

**Comment by portante** [comment]
What about removing the data points entirely, putting them into a bin to the side for the user to inspect in a table?  I am not entirely sure how to handle this either, at the moment, as there are supposed to be fixes to the tools, but we'll likely always have to handle such bad data.

**Comment by k-rister** [comment]
The part I struggle with is where to set the bar for when a datapoint should be flagged.  Additionally, I think I would have to create a methodology to reverse the flagging if the user suspects it's a false positive and I don't have a good picture in my mind what that should look like.  That said, unless I come up with something better this might be the approach I have to take.

**Comment by k-rister** [comment]
The jschart library uses D3 scales for the axes.  The scale has a boolean property called clamp -- if clamp is set to FALSE (the default) D3 will use it's interpolators to determine the appropriate coordinate to use for a given value.  The outlier's that some tools are generating appears to be breaking the interpolator and/or the rendering engine which utilizes the coordinates.  If clamp is set to TRUE then D3 behaves in an entirely different way -- any value that is outside of the domain will be assigned the coordinate that is associated with the nearest domain value.  If the domain is [0,1] and the value is -1 then D3 will treat the value as if it is 0.  Likewise, if the value is 2 then D3 will treat the value as if it is 1.  In jschart the clamp parameter is typically left disabled for X and Y axes scales in order to avoid mutations to the data.  For example, here is a single dataset that has been zoomed on the Y axis with Y axis clamp disabled:

![y axis clamp disabled](https://cloud.githubusercontent.com/assets/15911116/17669789/a913a178-62d4-11e6-9932-d81281164dca.png)

And here is the same dataset zoomed on the Y axis with Y axis clamp enabled:

![y axis clamp enabled](https://cloud.githubusercontent.com/assets/15911116/17669798/b7c5af72-62d4-11e6-97b2-8de6de59ee47.png)

In this particular example the effects are minimal, but if you examine the top of the chart closely you can see differences in the way the datapoints are rendered.  When there are fewer visible datapoints, such as when the X axis is also zoomed, the issue becomes more pronounced.  With clamp disabled:

![y axis clamp disabled zoomed](https://cloud.githubusercontent.com/assets/15911116/17670021/aa513464-62d5-11e6-8664-63a3948b38e3.png)

With clamp enabled:

![y axis clamp enabled zoomed](https://cloud.githubusercontent.com/assets/15911116/17669953/6473fc56-62d5-11e6-8caf-56314b464d77.png)

In order to avoid the mutations to the data clamp is disabled on both the X and Y axis for the chart axes scales.  Clamp is actually enabled for the scales that are used for the slider zoom controls in order to avoid some corner case issues that can arise due to small errors in the coordinate conversion process -- the slider control should never move past the domain minimum or maximum.  The same is not true for values which are no longer possible due to zooming.

However, the clamp parameter does have a possible application to the viewport.  By enabling clamp the outliers can no longer have coordinate values that are many orders of magnitude outside of the domain.  By enabling Y axis clamp and setting the domain to [0,13000] a usable chart with all datasets visible and functional is achieved:

![clamp-enabled-domain-set](https://cloud.githubusercontent.com/assets/15911116/17670231/8f118568-62d6-11e6-8129-798f12b7d5e1.png)

Even here the mutations associated with clamp are visible since the datasets with outliers appear to spike to the domain maximum of 13000 rather than extend beyond it.  However, the table with it's current value display shows otherwise since it is not affected by the mechanics of the rendering of the chart:

![clamp-enabled-domain-set-table](https://cloud.githubusercontent.com/assets/15911116/17670297/f1b66c7e-62d6-11e6-929b-76d5be810174.png)

Enabling clamp appears to be a viable way to handle the outliers in addition to controlling the domain.  Due to the nature in which data point values can be mutated I think it is not a good idea to turn clamping on always, but enabling the ability to dynamically control it along with the dynamic domain may be the solution to this problem.

**Comment by portante** [comment]
Seems like a good approach, let's do it and see how it plays out.

**Comment by k-rister** [comment]
I have made the changes to implement the fix for this issue.  Below are some screenshots that detail the changes.

First, a new row exists in the table, highlighted here with a red box:

![view-port-controls-1](https://cloud.githubusercontent.com/assets/15911116/18210240/2eb07ed8-70fd-11e6-8700-69471358cf0c.png)

When the the row is clicked it reveals the controls that have been added, again highlighted with a red box:

![view-port-controls-2](https://cloud.githubusercontent.com/assets/15911116/18210179/ed9b7650-70fc-11e6-868b-de5116af4024.png)

By changing the value for the Y Axis Max and clicking the "Update Axes" button the chart is much more legible:

![view-port-controls-3](https://cloud.githubusercontent.com/assets/15911116/18210186/f20bb632-70fc-11e6-8ba3-1b3d5bd119e7.png)

However, if you pay close attention you will see that the datasets with outliers have disappeared.  Enabling Y Axis Clamping (by clicking the checkbox, highlighted in red) fixes this and the datasets are once again visible, albeit distorted:

![view-port-controls-4](https://cloud.githubusercontent.com/assets/15911116/18210191/f7d557c6-70fc-11e6-8ace-28945c540887.png)
